src/main/java/com/zendesk/maxwell/producer/MaxwellKafkaProducer.java

```diff
@@ -25,60 +26,107 @@
+		                       Meter producedMessageMeter, Meter failedMessageMeter
+		) {
+			this.producedMessageCount = producedMessageCount;
+			this.failedMessageCount = failedMessageCount;
+			this.producedMessageMeter = producedMessageMeter;
+			this.failedMessageMeter = failedMessageMeter;
+			this.producer = producer;
+			this.fallbackTopic = fallbackTopic;
+			this.maxwell = maxwell;
+		}
+
+		public ProducerContext withoutFallbackTopic() {
+			return new ProducerContext(producer, maxwell, null,
+				producedMessageCount, failedMessageCount, producedMessageMeter, failedMessageMeter);
+		}
+	}
+
 	public static final Logger LOGGER = LoggerFactory.getLogger(MaxwellKafkaProducer.class);
 	private final AbstractAsyncProducer.CallbackCompleter cc;
 	private final Position position;
 	private final String json;
-	private final String key;
-	private final MaxwellContext context;
-
-	private Counter succeededMessageCount;
-	private Counter failedMessageCount;
-	private Meter succeededMessageMeter;
-	private Meter failedMessageMeter;
+	private final RowIdentity key;
+	private final ProducerContext context;
 
-	public KafkaCallback(AbstractAsyncProducer.CallbackCompleter cc, Position position, String key, String json,
-	                     Counter producedMessageCount, Counter failedMessageCount, Meter producedMessageMeter,
-	                     Meter failedMessageMeter, MaxwellContext context) {
+	public KafkaCallback(AbstractAsyncProducer.CallbackCompleter cc, Position position,
+	                     RowIdentity key, String json, ProducerContext context) {
 		this.cc = cc;
 		this.position = position;
 		this.key = key;
 		this.json = json;
-		this.succeededMessageCount = producedMessageCount;
-		this.failedMessageCount = failedMessageCount;
-		this.succeededMessageMeter = producedMessageMeter;
-		this.failedMessageMeter = failedMessageMeter;
 		this.context = context;
 	}
 
+	private KafkaCallback withContext(ProducerContext context) {
+		return new KafkaCallback(cc, position, key, json, context);
+	}
+
+	ProducerContext getContext() {
+		return context;
+	}
+
 	@Override
 	public void onCompletion(RecordMetadata md, Exception e) {
 		if ( e != null ) {
-			this.failedMessageCount.inc();
-			this.failedMessageMeter.mark();
+			context.failedMessageCount.inc();
+			context.failedMessageMeter.mark();
 
 			LOGGER.error(e.getClass().getSimpleName() + " @ " + position + " -- " + key);
 			LOGGER.error(e.getLocalizedMessage());
-			if ( e instanceof RecordTooLargeException ) {
-				LOGGER.error("Considering raising max.request.size broker-side.");
-			} else if (!this.context.getConfig().ignoreProducerError) {
-				this.context.terminate(e);
-				return;
+
+			boolean nonFatal = e instanceof RecordTooLargeException || context.maxwell.getConfig().ignoreProducerError;
+
+			if (nonFatal) {
+				if (context.fallbackTopic == null) {
+					cc.markCompleted();
+				} else {
+					publishFallback(md, e);
```

Comment:
We only need to publish for `RecordTooLargeException`.

Reply:
If you set `ignore_producer_errors` to `true`, and you set a fallback topic, this will publish every failed message to the fallback topic. This is easy to describe ("we either publish a successful message or a fallback one"), whereas only publishing a fallback record for _some_ errors and quietly dropping others seems confusing.

Reply:
I thought the purpose of this fallback topic is only for the messages which are too big for the original topic. Otherwise the messages destined for the original topic are ended up in the wrong topic (for example in the case of `TimeoutException`).

Reply:
👍

Reply:
I wonder if it wouldn't be better to pass the fallback message back through the worker queue?  Or are we too far removed from that?

Reply:
Good question. The current code _is_ threadsafe (since the producer is threadsafe, and we don't rely on other mutable state), though it lays a bit of a trap for introducing race conditions in the future.

It would mean you'd delay progress of in flight messages, which I don't like. The fallback item would go to the back of the queue, coming after up to 100 items which were after this one in the input partition. The queue has a fixed capacity, but I don't think it would ever get stuck since the worker doesn't wait for the callback completion.

It's also may just be too much of a pain, since that queue contains `RowMap`s, and this isn't much of a `RowMap`. You could introduce a new class encapsulating "rowmap or fallback record", but I'm not convinced it'd be worth the trouble.